### ISSUE TITLE
BUG: Null pointer dereference during unmount (busy inodes after unmount of zfs)

### ISSUE THREAD

**Issue by danielkza** [zfsonlinux/zfs#3314]
I'm running ZoL 0.6.4 on Fedora 21, kernel 3.18.7 (holding off on 3.19 due to the BTRFS mount hang bug). During shutdown I got this in my logs and had to reboot manually through SysRq.

```
Abr 18 23:15:39 daniel-pc kernel: VFS: Busy inodes after unmount of zfs. Self-destruct in 5 seconds.  Have a nice day...
Abr 18 23:15:43 daniel-pc kernel: VFS: Busy inodes after unmount of zfs. Self-destruct in 5 seconds.  Have a nice day...
Abr 18 23:15:44 daniel-pc kernel: BUG: unable to handle kernel NULL pointer dereference at 0000000000000050
Abr 18 23:15:44 daniel-pc kernel: IP: [<ffffffff81764e3e>] _raw_spin_lock+0xe/0x50
Abr 18 23:15:44 daniel-pc kernel: PGD 0 
Abr 18 23:15:44 daniel-pc kernel: Oops: 0002 [#1] SMP 
Abr 18 23:15:44 daniel-pc kernel: Modules linked in: rndis_host cdc_ether usbnet mii rfcomm ccm xt_CHECKSUM tun nf_conntrack_netbios_ns nf_conntrack_broadcast nf_conntrack_ftp ipt_MASQUERADE nf_nat_masquerade_ipv4 ip6t_rpfilter ip6t_REJECT nf_reject_ipv6 xt_conntrack ebtable_nat ebtable_broute bridge stp llc ebtable_filter ebtables ip6table_nat nf_conntrack_ipv6 nf_defrag_ipv6 nf_nat_ipv6 ip6table_mangle ip6table_security ip6table_raw ip6table_filter ip6_tables iptable_nat nf_conntrack_ipv4 nf_defrag_ipv4 nf_nat_ipv4 nf_nat nf_conntrack iptable_mangle iptable_security iptable_raw bnep nvidia(POE) vfat fat zfs(POE) zunicode(POE) fuse zavl(POE) zcommon(POE) znvpair(POE) intel_rapl eeepc_wmi asus_wmi sparse_keymap video x86_pkg_temp_thermal spl(OE) iTCO_wdt iTCO_vendor_support mxm_wmi kvm_intel snd_usb_audio kvm uas joydev
Abr 18 23:15:44 daniel-pc kernel:  usb_storage btusb snd_usbmidi_lib arc4 snd_rawmidi snd_hda_codec_hdmi crct10dif_pclmul bluetooth ath9k snd_hda_codec_realtek ath9k_common snd_hda_codec_generic crc32_pclmul ath9k_hw crc32c_intel ath ghash_clmulni_intel snd_hda_intel mac80211 snd_hda_controller snd_hda_codec cfg80211 snd_hwdep serio_raw snd_seq rfkill snd_seq_device i2c_i801 snd_pcm drm wmi lpc_ich mei_me mei snd_timer snd mfd_core soundcore shpchp binfmt_misc btrfs xor raid6_pq e1000e ptp pps_core [last unloaded: hwmon_vid]
Abr 18 23:15:44 daniel-pc kernel: CPU: 2 PID: 1374 Comm: zfs_iput_taskq/ Tainted: P           OE  3.18.7-200.fc21.x86_64 #1
Abr 18 23:15:44 daniel-pc kernel: Hardware name: System manufacturer System Product Name/P8P67 PRO, BIOS 2303 02/17/2012
Abr 18 23:15:44 daniel-pc kernel: task: ffff8804f0a409d0 ti: ffff88003e428000 task.ti: ffff88003e428000
Abr 18 23:15:44 daniel-pc kernel: RIP: 0010:[<ffffffff81764e3e>]  [<ffffffff81764e3e>] _raw_spin_lock+0xe/0x50
Abr 18 23:15:44 daniel-pc kernel: RSP: 0018:ffff88003e42bd28  EFLAGS: 00010292
Abr 18 23:15:44 daniel-pc kernel: RAX: 0000000000010000 RBX: ffff880441e1bd28 RCX: 0000000000000034
Abr 18 23:15:44 daniel-pc kernel: RDX: 0000000000000003 RSI: ffff88018426e2f8 RDI: 0000000000000050
Abr 18 23:15:44 daniel-pc kernel: RBP: ffff88003e42bd28 R08: 0000000000000202 R09: ffff8805027289d0
Abr 18 23:15:44 daniel-pc kernel: R10: 00000000ffffffff R11: ffff88035b410900 R12: ffff880441e1bd20
Abr 18 23:15:44 daniel-pc kernel: R13: 0000000000000000 R14: ffffffffa084fe40 R15: ffff8804ff084f00
Abr 18 23:15:44 daniel-pc kernel: FS:  0000000000000000(0000) GS:ffff88051f500000(0000) knlGS:0000000000000000
Abr 18 23:15:44 daniel-pc kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
Abr 18 23:15:44 daniel-pc kernel: CR2: 0000000000000050 CR3: 000000000cc11000 CR4: 00000000000407e0
Abr 18 23:15:44 daniel-pc kernel: Stack:
Abr 18 23:15:44 daniel-pc kernel:  ffff88003e42bd58 ffffffff8132035f 0000000000000000 ffff88018426e258
Abr 18 23:15:44 daniel-pc kernel:  ffff88018426e360 ffffffffa084fe40 ffff88003e42bd68 ffffffff81319bc6
Abr 18 23:15:44 daniel-pc kernel:  ffff88003e42bd88 ffffffff81232276 0000000000000000 ffff88018426e258
Abr 18 23:15:44 daniel-pc kernel: Call Trace:
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffff8132035f>] selinux_inode_free_security+0x2f/0x70
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffff81319bc6>] security_inode_free+0x16/0x20
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffff81232276>] __destroy_inode+0x26/0xe0
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffff81232356>] destroy_inode+0x26/0x70
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffff812324a7>] evict+0x107/0x190
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffff81232d25>] iput+0xf5/0x1a0
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffffa060b016>] taskq_thread+0x1b6/0x360 [spl]
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffff810c59a0>] ? wake_up_state+0x20/0x20
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffffa060ae60>] ? taskq_cancel_id+0x140/0x140 [spl]
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffff810b801a>] kthread+0xea/0x100
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffff810b7f30>] ? kthread_create_on_node+0x1b0/0x1b0
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffff8176537c>] ret_from_fork+0x7c/0xb0
Abr 18 23:15:44 daniel-pc kernel:  [<ffffffff810b7f30>] ? kthread_create_on_node+0x1b0/0x1b0
Abr 18 23:15:44 daniel-pc kernel: Code: 01 00 00 f0 0f c1 07 84 c0 75 02 c3 90 55 48 89 e5 e8 a7 bf 97 ff 5d c3 0f 1f 44 00 00 66 66 66 66 90 55 48 89 e5 b8 00 00 01 00 <f0> 0f c1 07 89 c2 c1 ea 10 66 39 c2 89 d1 75 02 5d c3 b8 00 80 
Abr 18 23:15:44 daniel-pc kernel: RIP  [<ffffffff81764e3e>] _raw_spin_lock+0xe/0x50
Abr 18 23:15:44 daniel-pc kernel:  RSP <ffff88003e42bd28>
Abr 18 23:15:44 daniel-pc kernel: CR2: 0000000000000050
Abr 18 23:15:44 daniel-pc kernel: ---[ end trace 428a11cf0f0db729 ]---
```

**Comment by behlendorf** [comment]
Possibly related to #3296.

**Comment by yesbox** [comment]
I'm frequently experiencing this issue on Centos 7.1 with ZoL 0.6.4.1 on shutdowns.

**Comment by behlendorf** [comment]
@yesbox commit f0da4d15082be83547e5370e4479d3c04b56451d should address this issue.  We're looking at applying this patch to the next point release.

**Comment by behlendorf** [comment]
This issue has been resolved in 0.6.5 and newer.
